brave-core/src/test/java/com/github/kristofa/brave/HttpSpanCollectorTest.java

```diff
@@ -0,0 +1,130 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import okio.Buffer;
+import okio.GzipSource;
+import org.junit.Rule;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class HttpSpanCollectorTest {
+
+  @Rule
+  public final MockWebServer server = new MockWebServer();
+
+  TestMetricsHander metrics = new TestMetricsHander();
+  // set flush interval to 0 so that tests can drive flushing explicitly
+  HttpSpanCollector collector = new HttpSpanCollector(server.url("").toString(), metrics, 0);
+
+  @Test
+  public void collectDoesntDoIO() throws Exception {
+    collector.collect(span(1L, "foo"));
+
+    assertThat(server.getRequestCount()).isZero();
+  }
+
+  @Test
+  public void collectIncrementsAcceptedMetrics() throws Exception {
+    collector.collect(span(1L, "foo"));
+
+    assertThat(metrics.acceptedSpans.get()).isEqualTo(1);
+    assertThat(metrics.droppedSpans.get()).isZero();
+  }
+
+  @Test
+  public void dropsWhenQueueIsFull() throws Exception {
+    for (int i = 0; i < 1001; i++)
+      collector.collect(span(1L, "foo"));
+
+    assertThat(metrics.acceptedSpans.get()).isEqualTo(1001);
+    assertThat(metrics.droppedSpans.get()).isEqualTo(1);
+  }
+
+  @Test
+  public void postsCompressedSpans() throws Exception {
+    server.enqueue(new MockResponse());
+
+    collector.collect(span(1L, "foo"));
+    collector.collect(span(2L, "bar"));
+
+    collector.flush(); // manually flush the spans
+
+    // Ensure a proper request was sent
+    RecordedRequest request = server.takeRequest();
+    assertThat(request.getRequestLine()).isEqualTo("POST /api/v1/spans HTTP/1.1");
+    assertThat(request.getHeader("Content-Type")).isEqualTo("application/x-thrift");
+    assertThat(request.getHeader("Content-Encoding")).isEqualTo("gzip");
+
+    // Now, let's read back the spans we sent!
+    List<io.zipkin.Span> zipkinSpans = Codec.THRIFT.readSpans(gunzip(request.getBody()));
+    assertThat(zipkinSpans).containsExactly(
+        zipkinSpan(1L, "foo"),
+        zipkinSpan(2L, "bar")
+    );
+  }
+
+  @Test
+  public void incrementsDroppedSpansWhenServerErrors() throws Exception {
+    server.enqueue(new MockResponse().setResponseCode(500));
+
+    collector.collect(span(1L, "foo"));
+    collector.collect(span(2L, "bar"));
+
+    collector.flush(); // manually flush the spans
+
+    assertThat(metrics.droppedSpans.get()).isEqualTo(2);
+  }
+
+  @Test
+  public void incrementsDroppedSpansWhenServerDisconnects() throws Exception {
+    server.enqueue(new MockResponse().setSocketPolicy(SocketPolicy.DISCONNECT_AFTER_REQUEST));
+
+    collector.collect(span(1L, "foo"));
+    collector.collect(span(2L, "bar"));
+
+    collector.flush(); // manually flush the spans
+
+    assertThat(metrics.droppedSpans.get()).isEqualTo(2);
+  }
+
+  class TestMetricsHander implements SpanCollectorMetricsHandler {
+
+    final AtomicInteger acceptedSpans = new AtomicInteger();
```

Comment:
these don't need to be atomic, yet, as I've not put any threaded tests in... that'll happen